proxy/stats_get_prediction.go

```diff
@@ -0,0 +1,361 @@
+package proxy
+
+// Precaches statistics/get calls if the opening sequence of calls is detected
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strings"
+)
+
+const StatsGetWorkers = 5
+
+type StatsGetTask struct {
+	request  string
+	response chan *http.Response
+}
+
+type StatsGetPrediction struct {
+	GGStriveAPIURL  string
+	loginPrefix     string
+	predictionState PredictionState
+	statsGetTasks   map[string]*StatsGetTask
+	client          *http.Client
+}
+
+type PredictionState int
+
+// Declare typed constants each with type of status
+const (
+	reset PredictionState = iota
+	get_env_called
+	login_parsed
+	sending_calls
+)
+
+type CachingResponseWriter struct {
+	w    http.ResponseWriter
+	buf  bytes.Buffer
+	code int
+}
+
+func (rw *CachingResponseWriter) Header() http.Header {
+	return rw.w.Header()
+}
+
+func (rw *CachingResponseWriter) WriteHeader(statusCode int) {
+	rw.w.WriteHeader(statusCode)
+	rw.code = statusCode
+}
+
+func (rw *CachingResponseWriter) Write(data []byte) (int, error) {
+	rw.buf.Write(data)
+	return rw.w.Write(data)
+}
+
+func (s *StatsGetPrediction) StatsGetStateHandler(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		path := r.URL.Path
+		if s.predictionState != reset &&
+			path != "/api/sys/get_env" &&
+			path != "/api/user/login" &&
+			path != "/api/statistics/get" {
+			if s.predictionState == sending_calls {
+				fmt.Println("Done looking up stats")
+			}
+			s.predictionState = reset
+		}
+
+		if path == "/api/sys/get_env" || path == "/api/user/login" {
+			wrappedWriter := CachingResponseWriter{w: w}
+			next.ServeHTTP(&wrappedWriter, r)
+
+			if path == "/api/sys/get_env" && wrappedWriter.code < 400 {
+				s.predictionState = get_env_called
+			} else if path == "/api/user/login" &&
+				wrappedWriter.code < 400 &&
+				s.predictionState == get_env_called {
+				login := make([]byte, wrappedWriter.buf.Len())
+
+				wrappedWriter.buf.Read(login)
+				s.ParseLoginPrefix(login)
+				s.predictionState = login_parsed
+			}
+		} else {
+			next.ServeHTTP(w, r)
+		}
+	})
+
+}
+
+// Proxy getstats
+func (s *StatsGetPrediction) HandleGetStats(w http.ResponseWriter, r *http.Request) bool {
+	if len(s.loginPrefix) > 0 && s.predictionState == login_parsed {
+		s.AsyncGetStats()
+		s.predictionState = sending_calls
+	}
+	if len(s.loginPrefix) > 0 && s.predictionState == sending_calls {
+		bodyBytes, _ := ioutil.ReadAll(r.Body)
+		r.Body.Close() //  must close
+		r.Body = ioutil.NopCloser(bytes.NewBuffer(bodyBytes))
```

Comment:
Would be good to explain why you're resetting `r.Body` here, since for anyone looking at only this function, it doesn't make any sense.

I understand it's used by `HandleCatchall()` after `HandleGetStats()` returns, but it's not obvious in context.